processing-service-core/api/models/tenant.js

```diff
@@ -0,0 +1,340 @@
+                                for (var key in tenantDefaultConfig) {
+                                    effectiveConfig.value = tenantDefaultConfig[key];
+                                    effectiveConfig.valueType = "DEFAULT";
+                                    tenantConfig.resolve(effectiveConfig);
+                                }
+                            })
+                            .fail(function (errorResponse) {
+                                tenantConfig.reject(errorResponse);
+                            })
+                            .done();
+                } else {
+                    tenantConfig.reject(errorResponse);
+                }
+            })
+            .done();
+    return tenantConfig.promise;
+}
+
+/**
+ * Retrieve effective tenant specific config.
+ * Returns the effective tenant specific config for this key. That can be either a custom value specified for the tenant or the default.
+ *
+ * @param {String} tenantId String The ID tenant.
+ * returns effectiveTenantConfigs
+ **/
+function getEffectiveTenantConfigs(tenantId) {
+
+    var tenantConfig = Q.defer();
+    logger.debug("Retrieving efective tenant config for tenant " + tenantId);
+    var effectiveConfigs = {};
+    effectiveConfigs.tenantId = tenantId;
+    effectiveConfigs.configurations = [];
+    tenantConfigStoreModel.getTenantConfigs(tenantId)
+            .then(function (tenantConfigurations) {
+                var tenantConfigMap = createMap(tenantConfigurations);
+                for (var key in tenantConfigMap) {
+                    var config = {
+                        tenantId: tenantId,
+                        key: key,
+                        value: tenantConfigMap[key],
+                        valueType: "CUSTOM"
+                    };
+                    effectiveConfigs.configurations.push(config);
+                }
+            })
+            .fail(function (errorResponse) {
+                //Not throwing the error as this can happen when a tenant has not set any custom configurations.
+                logger.debug("Failed to find any tenant specific configs for tenant " + tenantId);
+            })
+            .done();
+
+    globalConfigStoreModel.getGlobalConfigs(tenantId)
+            .then(function (globalConfig) {
+                for (var item in globalConfig) {
+                    for (var key in globalConfig[item]) {
+                        if (containsKey(effectiveConfigs.configurations, key) === false) {
+                            var config = {
+                                tenantId: tenantId,
+                                key: key,
+                                value: globalConfig[item][key],
+                                valueType: "DEFAULT"
+                            };
+                            effectiveConfigs.configurations.push(config);
+                        }
+                    }
+                    tenantConfig.resolve(effectiveConfigs);
```

Comment:
There's nothing preventing this line being reached before getTenantConfigs has returned and performed its logic. You should create both promises and wait on both finishing before calling this resolve.